CMakeLists.txt

```diff
@@ -413,37 +413,45 @@ if(USE_CUDA)
   set_target_properties(lightgbm PROPERTIES CUDA_RESOLVE_DEVICE_SYMBOLS ON)
   TARGET_LINK_LIBRARIES(
     lightgbm
+    PRIVATE
     ${histograms}
   )
   set_target_properties(_lightgbm PROPERTIES CUDA_RESOLVE_DEVICE_SYMBOLS ON)
   TARGET_LINK_LIBRARIES(
     _lightgbm
+    PRIVATE
     ${histograms}
   )
 endif(USE_CUDA)
 
 if(USE_HDFS)
-  TARGET_LINK_LIBRARIES(lightgbm ${HDFS_CXX_LIBRARIES})
-  TARGET_LINK_LIBRARIES(_lightgbm ${HDFS_CXX_LIBRARIES})
+  TARGET_LINK_LIBRARIES(lightgbm PRIVATE ${HDFS_CXX_LIBRARIES})
+  TARGET_LINK_LIBRARIES(_lightgbm PRIVATE ${HDFS_CXX_LIBRARIES})
 endif(USE_HDFS)
 
 if(WIN32)
     if(MINGW OR CYGWIN)
-      TARGET_LINK_LIBRARIES(lightgbm Ws2_32)
-      TARGET_LINK_LIBRARIES(_lightgbm Ws2_32)
-      TARGET_LINK_LIBRARIES(lightgbm IPHLPAPI)
-      TARGET_LINK_LIBRARIES(_lightgbm IPHLPAPI)
+      TARGET_LINK_LIBRARIES(lightgbm PRIVATE Ws2_32)
+      TARGET_LINK_LIBRARIES(_lightgbm PRIVATE Ws2_32)
+      TARGET_LINK_LIBRARIES(lightgbm PRIVATE IPHLPAPI)
+      TARGET_LINK_LIBRARIES(_lightgbm PRIVATE IPHLPAPI)
     endif(MINGW OR CYGWIN)
 endif(WIN32)
 
 if(__BUILD_FOR_R)
   if(MSVC)
-    TARGET_LINK_LIBRARIES(_lightgbm ${LIBR_MSVC_CORE_LIBRARY})
+    TARGET_LINK_LIBRARIES(_lightgbm PRIVATE ${LIBR_MSVC_CORE_LIBRARY})
   else()
-    TARGET_LINK_LIBRARIES(_lightgbm ${LIBR_CORE_LIBRARY})
+    TARGET_LINK_LIBRARIES(_lightgbm PRIVATE ${LIBR_CORE_LIBRARY})
```

Comment:
why are your proposing adding  all these `PRIVATE` annotations?

Reply:
I think I had to add PRIVATE to fmt that to silence some warnings, and when you add that you have to add them everywhere because you can't mix old-style and new-style target_link_library() calls in the same CMake, but I'll try removing it again after merging your patch.

Reply:
Ok, just tried removing it, I cannot, fmt needs this to work in header-only mode otherwise, duplicate function definitions appear all over and break the compilation.
Then due to lack of compatibility between new and old-style target_link_libraries, one must also specify the link type for each  library.

Reply:
There are some issues with omp on MacOS though, might be related to this, and as explained here: https://cmake.org/pipermail/cmake/2016-May/063400.html

Maybe OMP it has to be PUBLIC in this case. I'll try it out.

Reply:
for the CRAN changes, just moving the `fmt` headers into the right place and passing `-DFMT_HEADER_ONLY` was enough to successfully compile and run `LightGBM` + `fmt`: https://github.com/feedzai/LightGBM/pull/7/files

could the same be done here? Basically treating `fmt` as a part of LightGBM, instead of a separate library that we compile and link against.

I ask because I'm worried about the implication of marking all those things `PRIVATE`. I'm not sure if that will cause issues for projects that link against lib_lightgbm, like https://github.com/microsoft/LightGBM#external-unofficial-repositories.

Reply:
can you please answer this when you have time @AlbertoEAF

Reply:
Hi there James, of course,

Great reminder, I will remove them tomorrow, no reason to have them anymore ;)